src/accumulator/proof.rs

```diff
@@ -249,15 +250,154 @@ impl Proof {
     pub fn update(
         self,
         cached_hashes: Vec<sha256::Hash>,
+        add_hashes: Vec<sha256::Hash>,
         block_targets: Vec<u64>,
+        remembers: Vec<u64>,
         update_data: UpdateData,
     ) -> Result<(Proof, Vec<sha256::Hash>), String> {
-        self.update_proof_remove(
+        let (proof_after_deletion, cached_hashes) = self.update_proof_remove(
             block_targets,
             cached_hashes.clone(),
             update_data.new_del,
             update_data.prev_num_leaves,
-        )
+        )?;
+
+        let data_after_addition = proof_after_deletion.update_proof_add(
+            add_hashes,
+            cached_hashes.clone(),
+            remembers,
+            update_data.new_add,
+            update_data.prev_num_leaves,
+            update_data.to_destroy,
+        )?;
+
+        Ok(data_after_addition)
+    }
+    fn update_proof_add(
+        self,
+        adds: Vec<sha256::Hash>,
+        cached_del_hashes: Vec<sha256::Hash>,
+        remembers: Vec<u64>,
+        new_nodes: Vec<(u64, sha256::Hash)>,
+        before_num_leaves: u64,
+        to_destroy: Vec<u64>,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        // Combine the hashes with the targets.
+        let orig_targets_with_hash: Vec<(u64, sha256::Hash)> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(cached_del_hashes.into_iter())
+            .collect();
+
+        // Attach positions to the proof.
+        let proof_pos = get_proof_positions(
+            &self.targets,
+            before_num_leaves,
+            util::tree_rows(before_num_leaves),
+        );
+        let proof_with_pos = proof_pos
+            .clone()
+            .into_iter()
+            .zip(self.hashes.clone())
+            .collect();
+
+        // Remap the positions if we moved up a after the addition row.
```

Comment:
nit: Could you change the comment to:
`Remap the positions if we moved up a row after the addition.`

Just making things grammatically correct